カレンダーのプラクティスを提出

mariri777 opened the pull request:
1回目の提出をします。
よろしくお願いします。

mariri777 committed:
カレンダーのプラクティスを提出

mariri777 committed:
カレンダーのプラクティスを提出

yoshitsugu commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,40 @@
+  if (1970..2100).cover?(year)
+    options[:year] = year
+  else
+    puts "指定範囲外です"
```
範囲チェックできているのはよいですね :+1: 
月と年が同じメッセージだとどちらが範囲外だったのかわかりづらいので、「月指定が~」などにしたほうがいいかなと思いました。

yoshitsugu commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,40 @@
+first_day.wday.times { print "   " }
+
+(first_day..last_day).each do |date|
+  print format("%2d ", date.day)
```
これでもよいのですが、 `String#rjust` `String#ljust` というもうちょっと人間が読みやすいメソッドがあるので、こちらも見てみてください。
https://docs.ruby-lang.org/ja/latest/method/String/i/rjust.html
https://docs.ruby-lang.org/ja/latest/method/String/i/ljust.html

yoshitsugu commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,40 @@
+
+(first_day..last_day).each do |date|
+  print format("%2d ", date.day)
+  if date.wday == 6
```
Dateオブジェクトには `saturday?` というメソッドもあります。
https://docs.ruby-lang.org/ja/latest/method/Date/i/saturday=3f.html
このように、Rubyには便利メソッドが標準でたくさん用意されているので、「これ他の人も同じ問題に困ってそうだな？」というのがあればまずはリファレンスマニュアルを調べてもらうと該当のメソッドが見つかるかもしれません。

yoshitsugu commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,40 @@
+(first_day..last_day).each do |date|
+  print format("%2d ", date.day)
+  if date.wday == 6
+    puts ""
```
改行のみ出力したいときは引数なしで大丈夫です。
https://docs.ruby-lang.org/ja/latest/method/Kernel/m/puts.html

リファレンスマニュアルにも

> 引数がなければ改行のみを出力します。 

とあります。

yoshitsugu replied:
出力の最後にも改行を出力するようにしてください。環境によっては続けてプロンプトが表示されてしまいます。
```
[yoshitsugu]$ ./cal.rb
      3月 2024
日 月 火 水 木 金 土
                1  2
 3  4  5  6  7  8  9
10 11 12 13 14 15 16
17 18 19 20 21 22 23
24 25 26 27 28 29 30
31 [yoshitsugu]$
```

yoshitsugu reviewed (commented):
コメントしました！

mariri777 committed:
2回目の提出

1.月指定、年指定の文言を変更
2.`rjust`を使用し文字列を整形
3.`saturday?`メソッドを使用
4.引数なしで改行を出力
5.出力の最後に改行を追加

mariri777 committed:
2回目の提出

1.月指定、年指定の文言を変更
2.`rjust`を使用し文字列を整形
3.`saturday?`メソッドを使用
4.引数なしで改行を出力
5.出力の最後に改行を追加

mariri777 commented:
@yoshitsugu 
レビューしていただきありがとうございました。
ご助言いただいた箇所を修正いたしました。
再度ご確認をよろしくお願いいたします。

yoshitsugu commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,42 @@
+first_day.wday.times { print "   "}
+
+(first_day..last_day).each do |date|
+  print date.day.to_s.rjust(3)
```
日曜日は左側に空白が入らないように考えてみてください。

```
      3月 2024
日 月 火 水 木 金 土
                 1  2
  3  4  5  6  7  8  9
 10 11 12 13 14 15 16
 17 18 19 20 21 22 23
 24 25 26 27 28 29 30
 31
```
ではなく
```
      3月 2024
日 月 火 水 木 金 土
                1  2
 3  4  5  6  7  8  9
10 11 12 13 14 15 16
17 18 19 20 21 22 23
24 25 26 27 28 29 30
31
```

yoshitsugu reviewed (commented):
コメントしました！

mariri777 committed:
3回目の提出

1.後置ifに修正
2.左側に空白が入らないよう整形3回目の提出

1.後置ifに修正
2.左側に空白が入らないよう整形3回目の提出

1.後置ifに修正
2.左側に空白が入らないよう整形

mariri777 committed:
Merge branch 'calendar' of https://github.com/mariri777/ruby-practices into calendar

mariri777 commented:
@yoshitsugu 
ご確認ありがとうございました。
再度ご確認をお願いいたします。

レビューいただいたものの他に変更点があります。
rubocopを使用したら
`cal.rb:13:10: C: [Correctable] Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.`
と出てきたので、ダブルクォーテーションのところはシングルクォーテーションに変更しています。

また、課題には関係がないのですが、コミットメッセージがなぜか3回繰り返して登録されていました。。
1回分しか登録していないはずなのですが、これは私がどこかで操作を誤ったのでしょうか？
もしよろしければご教示いただけますと幸いです。

どうぞよろしくお願いいたします。

yoshitsugu commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,40 @@
+  if (1970..2100).cover?(year)
+    options[:year] = year
+  else
+    puts '年の指定が範囲外です'
```
エラーメッセージは標準エラー出力に表示することが一般的に想定されます。
また、エラーのときは `exit(1)` や `abort` などを使って、終了ステータスを1(以上)にして、終了することが一般的です。
https://docs.ruby-lang.org/ja/latest/method/Kernel/m/exit.html
https://docs.ruby-lang.org/ja/latest/method/Kernel/m/abort.html

標準エラー出力、終了ステータスなどがわからなければ調べてみてください〜

yoshitsugu replied:
標準エラー出力への出力ができてないようです〜

yoshitsugu reviewed (commented):
コメントしました！

> また、課題には関係がないのですが、コミットメッセージがなぜか3回繰り返して登録されていました。。
> 1回分しか登録していないはずなのですが、これは私がどこかで操作を誤ったのでしょうか？
> もしよろしければご教示いただけますと幸いです。

どこをどうやってコミットしたのかがわかないとなんとも言えないのですが、基本的にはご自身の誤操作を疑ったほうがよいかと思います。

mariri777 committed:
4回目の提出

1.無効な入力を処理するため、`exit`ステートメントを追加

mariri777 commented:
@yoshitsugu 
レビューいただきありがとうございます。
度々申し訳ございませんが、再度ご確認をお願いいたします。

> どこをどうやってコミットしたのかがわかないとなんとも言えないのですが、基本的にはご自身の誤操作を疑ったほうがよいかと思います。
抽象的な質問となり申し訳ございません。
通常通りコミット後Vimが開きコミットメッセージを入力したのですが、どこかで操作を誤ったのだと思います。
今後このようなことがないようにもう少し調べてみます。
ご回答いただきありがとうございました。

yoshitsugu reviewed (commented):
コメントしました！

Gitのコミットについていうと、コミットの単位はまとめて一つではなく、キリのいい単位で細かく積んでいくほうが一般的によいとされています。なぜかというとその修正だけ戻したいときに困るから、などの理由です。次回から気をつけてみてください。
参考: https://bootcamp.fjord.jp/pages/make-your-commit-message-better

mariri777 committed:
標準エラー出力を追加

mariri777 commented:
@yoshitsugu 
標準出力エラーを調べて追加しました。
再度ご確認をお願いいたします。

また、コミットの単位についてご教示いただきありがとうございます。
今後気をつけていきたいと思います。

yoshitsugu reviewed (approved):
OKです〜

mariri777 commented:
@yoshitsugu 
ご確認いただきありがとうございました！
